module/mempool/herocache/backdata/heropool/pool.go

```diff
@@ -262,9 +271,24 @@ func (p Pool) isInvalidated(sliceIndex EIndex) bool {
 	return true
 }
 
+// a helper method that allows to get an adress fo the state form the state type.
+func (p *Pool) getStateFromType(stateType StateType) *state {
+	var s *state = nil
+	switch stateType {
+	case stateFree:
+		s = &p.free
+	case stateUsed:
+		s = &p.used
+	default:
+		panic("Unknown state type")
+	}
+	return s
+}
```

Comment:
Indeed, the creation of a heap reference `var s *state` within this method contradicts one of the design principles of `HeroCache` that I had overlooked earlier. To rectify this, please incorporate the method's logic directly within `removeEntity` and `appendEntity` without the necessity for an additional method. Furthermore, for more descriptive error messaging during a panic, employ `panic(fmt.Sprintf("unknown state type: %s", s))`.

Reply:
thx, will implement

Reply:
A summary, of a current update 
1. done , pls check, thanks you for giving examples in code.
2. done, pls take a look.
3. done, probabilities are configurable, so we might want to fiddle with them or even have an array of probabilities to test
4. Thank you for providing the test for benchmark. I got similar results :
master - 
ok      github.com/onflow/flow-go/module/mempool/stdmap 549.314s
current branch -
ok      github.com/onflow/flow-go/module/mempool/stdmap 557.312s

there is 1% difference.
Test time is now under one second. I guess reading from file was a bottleneck.

Reply:
Dear @yhassanzadeh13, I think I addressed your previous comments. I am still in doubt about how one has to log the seed in order to make it trackable on CI but other than that I think things are addressed. Please, take another look when you have time.

Reply:
Thanks @kozlovb for applying the comments. 
Regarding:
1. I reviewed a test, it looks good in general. Added a few minor comments. Please address them, and then the test is ready. 
2. Thanks, the logic looks good for the PR considering that we are moving to a multi-state logic soon. 
3. Looks good, thanks for applying that. 
4. Thanks for the benchmark, it looks good, please also add the heap info prior and post garbage collection in both your implementation and `master`, you can just run [this method]((https://github.com/onflow/flow-go/blob/master/module/mempool/stdmap/backDataHeapBenchmark_test.go#L45-L65)) on both branches and paste the numbers here.

Reply:
Thank you, @kozlovb, for your efforts in addressing the comments.

Regarding your updates:

1. I've reviewed the modified test, and overall it appears to be in a good state. However, there are a few minor points that I've noted. Please consider addressing these, after which the test should be fully prepared.

2. The logic for the PR seems sound, especially considering our upcoming transition to a multi-state logic. I appreciate your foresight in this matter.

3. The implemented changes are satisfactory. Thanks for your prompt response.

4. I appreciate the benchmark data provided. It's quite informative, but I would also appreciate it if you could include heap information before and after garbage collection for both your implementation and the `master` branch. You can simply execute [this method](https://github.com/onflow/flow-go/blob/master/module/mempool/stdmap/backDataHeapBenchmark_test.go#L45-L65) on both branches and include the results here.

Regarding the CI, I believe there has been a bit of a misunderstanding. While I appreciate your effort to log the seed, no further action is required on this front. My initial comment pertained to generating test cases dynamically during test execution instead of reading them from an external file. This approach would necessitate configuring our CI to ensure that the test case generation occurs before the main test process so that the testing can be performed with fresh data samples. However, your recent changes have addressed these concerns, so there's no issue remaining. Thanks again for your responsiveness.

Reply:
@yhassanzadeh13 thank you for your comments and suggestions. I hope I didn't miss anything.
Please take another look at this PR.

Here are the outputs of the benchmark for this branch - 
unning tool: /usr/local/go/bin/go test -benchmem -run=^$ -bench ^BenchmarkArrayBackDataLRU$ github.com/onflow/flow-go/module/mempool/stdmap

{"level":"info","interaction_time":442200.414215,"time":"2023-07-12T23:07:47Z","message":"adding elements done"}
{"level":"info","gc-elapsed-time":15.296439469,"time":"2023-07-12T23:08:02Z","message":"garbage collection done"}
goos: darwin
goarch: amd64
pkg: github.com/onflow/flow-go/module/mempool/stdmap
cpu: Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
BenchmarkArrayBackDataLRU-12    	       1	566091262999 ns/op	15210315016 B/op	600000648 allocs/op
PASS
ok  	github.com/onflow/flow-go/module/mempool/stdmap	568.669s

and here is for master - 

Running tool: /usr/local/go/bin/go test -benchmem -run=^$ -bench ^BenchmarkArrayBackDataLRU$ github.com/onflow/flow-go/module/mempool/stdmap

{"level":"info","interaction_time":448399.164792,"time":"2023-07-12T23:29:58Z","message":"adding elements done"}
{"level":"info","gc-elapsed-time":8.190857803,"time":"2023-07-12T23:30:06Z","message":"garbage collection done"}
goos: darwin
goarch: amd64
pkg: github.com/onflow/flow-go/module/mempool/stdmap
cpu: Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
BenchmarkArrayBackDataLRU-12    	       1	558280645412 ns/op	15210315008 B/op	600000647 allocs/op
PASS
ok  	github.com/onflow/flow-go/module/mempool/stdmap	560.739s

Reply:
Thanks for sharing the results @kozlovb. The first look was quite concerning as the `gc-elapsed-time` was doubled on your branch compared to `master` which would typically signal a performance regression. However, I run the comparison locally in isolation and the result is quite acceptable. There is a regression on your branch but we can ignore it for now. Please post the following result on the PR description as well. 

your branch: 
```
{"level":"info","interaction_time":204036.611041,"time":"2023-07-13T20:58:46Z","message":"adding elements done"}
{"level":"info",".Alloc":15212417928,".AllocObjects":500008414,".TotalAlloc":15218387072,".NumGC":5,"time":"2023-07-13T20:58:46Z","message":"heap allocation digest"}
{"level":"info","gc-elapsed-time":0.547496792,"time":"2023-07-13T20:58:46Z","message":"garbage collection done"}
{"level":"info",".Alloc":2479496,".AllocObjects":8130,".TotalAlloc":15218389512,".NumGC":6,"time":"2023-07-13T20:58:46Z","message":"heap allocation digest"}
BenchmarkArrayBackDataLRU-10    	       1	239270562792 ns/op
PASS
```

`flow-go/master`

```
{"level":"info","interaction_time":197588.858917,"time":"2023-07-13T20:50:06Z","message":"adding elements done"}
{"level":"info",".Alloc":15212329024,".AllocObjects":500007132,".TotalAlloc":15218001832,".NumGC":4,"time":"2023-07-13T20:50:06Z","message":"heap allocation digest"}
{"level":"info","gc-elapsed-time":0.513674208,"time":"2023-07-13T20:50:06Z","message":"garbage collection done"}
{"level":"info",".Alloc":2408336,".AllocObjects":6876,".TotalAlloc":15218006512,".NumGC":5,"time":"2023-07-13T20:50:06Z","message":"heap allocation digest"}
BenchmarkArrayBackDataLRU-10    	       1	233279900958 ns/op
```

Reply:
Thank you for taking your time to run it on a your local machine. An interesting fluctuation in the ```gc-elapsed-time```. Sure, adding it now.